### ISSUE TITLE
Some warning when using pretrain model

### ISSUE THREAD

**Issue by lucky9-cyou** [WHU-USI3DV/SparseDC#2]
When I use pretrained model and eval_nyu.sh, the following warning will appear:

```bash
size mismatch for patch_embed1.proj.weight: copying a param with shape torch.Size([64, 3, 7, 7]) from checkpoint, the shape in current model is torch.Size([64, 64, 7, 7]).
size mismatch for block1.0.norm2.weight: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([128]).
size mismatch for block1.0.norm2.bias: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([128]).
size mismatch for block1.0.mlp.fc1.weight: copying a param with shape torch.Size([512, 64]) from checkpoint, the shape in current model is torch.Size([512, 128]).
size mismatch for block1.1.norm2.weight: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([128]).
size mismatch for block1.1.norm2.bias: copying a param with shape torch.Size([64]) from checkpoint, the shape in current model is torch.Size([128]).
size mismatch for block1.1.mlp.fc1.weight: copying a param with shape torch.Size([512, 64]) from checkpoint, the shape in current model is torch.Size([512, 128]).
size mismatch for block2.0.norm2.weight: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([256]).
size mismatch for block2.0.norm2.bias: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([256]).
size mismatch for block2.0.mlp.fc1.weight: copying a param with shape torch.Size([1024, 128]) from checkpoint, the shape in current model is torch.Size([1024, 256]).
size mismatch for block2.1.norm2.weight: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([256]).
size mismatch for block2.1.norm2.bias: copying a param with shape torch.Size([128]) from checkpoint, the shape in current model is torch.Size([256]).
size mismatch for block2.1.mlp.fc1.weight: copying a param with shape torch.Size([1024, 128]) from checkpoint, the shape in current model is torch.Size([1024, 256]).
size mismatch for block3.0.norm2.weight: copying a param with shape torch.Size([320]) from checkpoint, the shape in current model is torch.Size([640]).
size mismatch for block3.0.norm2.bias: copying a param with shape torch.Size([320]) from checkpoint, the shape in current model is torch.Size([640]).
size mismatch for block3.0.mlp.fc1.weight: copying a param with shape torch.Size([1280, 320]) from checkpoint, the shape in current model is torch.Size([1280, 640]).
size mismatch for block3.1.norm2.weight: copying a param with shape torch.Size([320]) from checkpoint, the shape in current model is torch.Size([640]).
size mismatch for block3.1.norm2.bias: copying a param with shape torch.Size([320]) from checkpoint, the shape in current model is torch.Size([640]).
size mismatch for block3.1.mlp.fc1.weight: copying a param with shape torch.Size([1280, 320]) from checkpoint, the shape in current model is torch.Size([1280, 640]).
size mismatch for block4.0.norm2.weight: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([1024]).
size mismatch for block4.0.norm2.bias: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([1024]).
size mismatch for block4.0.mlp.fc1.weight: copying a param with shape torch.Size([2048, 512]) from checkpoint, the shape in current model is torch.Size([2048, 1024]).
size mismatch for block4.1.norm2.weight: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([1024]).
size mismatch for block4.1.norm2.bias: copying a param with shape torch.Size([512]) from checkpoint, the shape in current model is torch.Size([1024]).
size mismatch for block4.1.mlp.fc1.weight: copying a param with shape torch.Size([2048, 512]) from checkpoint, the shape in current model is torch.Size([2048, 1024]).
unexpected key in source state_dict: head.weight, head.bias, patch_embed1.norm.weight, patch_embed1.norm.bias, block1.0.attn.norm.weight, block1.0.attn.norm.bias, block1.1.attn.norm.weight, block1.1.attn.norm.bias, patch_embed2.norm.weight, patch_embed2.norm.bias, block2.0.attn.norm.weight, block2.0.attn.norm.bias, block2.1.attn.norm.weight, block2.1.attn.norm.bias, patch_embed3.norm.weight, patch_embed3.norm.bias, block3.0.attn.norm.weight, block3.0.attn.norm.bias, block3.1.attn.norm.weight, block3.1.attn.norm.bias, patch_embed4.norm.weight, patch_embed4.norm.bias

missing keys in source state_dict: patch_embed1.norm.norm.weight, patch_embed1.norm.norm.bias, patch_embed1.norm.norm.running_mean, patch_embed1.norm.norm.running_var, block1.0.attn.norm.norm.weight, block1.0.attn.norm.norm.bias, block1.0.attn.norm.norm.running_mean, block1.0.attn.norm.norm.running_var, block1.0.mlp.norm.norm.weight, block1.0.mlp.norm.norm.bias, block1.0.mlp.norm.norm.running_mean, block1.0.mlp.norm.norm.running_var, block1.1.attn.norm.norm.weight, block1.1.attn.norm.norm.bias, block1.1.attn.norm.norm.running_mean, block1.1.attn.norm.norm.running_var, block1.1.mlp.norm.norm.weight, block1.1.mlp.norm.norm.bias, block1.1.mlp.norm.norm.running_mean, block1.1.mlp.norm.norm.running_var, patch_embed2.norm.norm.weight, patch_embed2.norm.norm.bias, patch_embed2.norm.norm.running_mean, patch_embed2.norm.norm.running_var, block2.0.attn.norm.norm.weight, block2.0.attn.norm.norm.bias, block2.0.attn.norm.norm.running_mean, block2.0.attn.norm.norm.running_var, block2.0.mlp.norm.norm.weight, block2.0.mlp.norm.norm.bias, block2.0.mlp.norm.norm.running_mean, block2.0.mlp.norm.norm.running_var, block2.1.attn.norm.norm.weight, block2.1.attn.norm.norm.bias, block2.1.attn.norm.norm.running_mean, block2.1.attn.norm.norm.running_var, block2.1.mlp.norm.norm.weight, block2.1.mlp.norm.norm.bias, block2.1.mlp.norm.norm.running_mean, block2.1.mlp.norm.norm.running_var, patch_embed3.norm.norm.weight, patch_embed3.norm.norm.bias, patch_embed3.norm.norm.running_mean, patch_embed3.norm.norm.running_var, block3.0.attn.norm.norm.weight, block3.0.attn.norm.norm.bias, block3.0.attn.norm.norm.running_mean, block3.0.attn.norm.norm.running_var, block3.0.mlp.norm.norm.weight, block3.0.mlp.norm.norm.bias, block3.0.mlp.norm.norm.running_mean, block3.0.mlp.norm.norm.running_var, block3.1.attn.norm.norm.weight, block3.1.attn.norm.norm.bias, block3.1.attn.norm.norm.running_mean, block3.1.attn.norm.norm.running_var, block3.1.mlp.norm.norm.weight, block3.1.mlp.norm.norm.bias, block3.1.mlp.norm.norm.running_mean, block3.1.mlp.norm.norm.running_var, patch_embed4.norm.norm.weight, patch_embed4.norm.norm.bias, patch_embed4.norm.norm.running_mean, patch_embed4.norm.norm.running_var, block4.0.mlp.norm.norm.weight, block4.0.mlp.norm.norm.bias, block4.0.mlp.norm.norm.running_mean, block4.0.mlp.norm.norm.running_var, block4.1.mlp.norm.norm.weight, block4.1.mlp.norm.norm.bias, block4.1.mlp.norm.norm.running_mean, block4.1.mlp.norm.norm.running_var
```

I want to ask, is this a normal phenomenon?

**Comment by lucky9-cyou** [comment]
But the evaluation result seems to be right. The following results are from the nyu with shift grid sample:
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test/D^1          │    0.8964150547981262     │
│         test/D^2          │    0.9770693778991699     │
│         test/D^3          │     0.994140088558197     │
│         test/MAE          │    0.23583249747753143    │
│         test/REL          │    0.08584385365247726    │
│         test/RMSE         │    0.3812095522880554     │
│         test/iMAE         │   0.041311733424663544    │
│        test/iRMSE         │    0.06896471977233887    │
└───────────────────────────┴───────────────────────────┘
RMSE : 0.3736±0.0035
MAE  : 0.2331±0.0020
iRMSE: 0.0680±0.0009
iMAE : 0.0409±0.0005
REL  : 0.0852±0.0009
D^1  : 0.8988±0.0030
D^2  : 0.9782±0.0011
D^3  : 0.9947±0.0005

**Comment by lucky9-cyou** [comment]
I found that this should be related to batch normalization. Thank you for your excellent work.

**Comment by lucky9-cyou** [comment]
The parameters of BN is still 1/0 after load pretrained model
https://github.com/WHU-USI3DV/SparseDC/blob/608cd1914fee3d816d25fdb1a9862058af759836/src/models/backbones/pvtv2.py#L519

**Comment by TouqeerAhmad** [comment]
I am seeing the same mismatch issue. Sharing the screenshot for clarity. 
![Uploading Screenshot 2024-01-19 at 10.28.07 AM.png…]()

**Comment by chenlongwhu** [comment]
Thank you for your attention. This is a  normal phenomenon since the original PVT directly encodes RGB images (H, W, 3), while in our case, we encode the features (H, W, C) from the SFFM module and input them into the PVT. As a result, we have made some modifications to the convolutional kernel sizes.  I will optimize my code to avoid these warnings at a later time. Thank you for your comments.

**Comment by lucky9-cyou** [comment]
> Thank you for your attention. This is a normal phenomenon since the original PVT directly encodes RGB images (H, W, 3), while in our case, we encode the features (H, W, C) from the SFFM module and input them into the PVT. As a result, we have made some modifications to the convolutional kernel sizes. I will optimize my code to avoid these warnings at a later time. Thank you for your comments.

I noticed that. Do you mean that the input for the PVT pre-trained model you are using does not conform to your design specifications? What I'm more curious about is why this code still functions properly when not all the weights are from pre-training.

**Comment by chenlongwhu** [comment]
@lucky9-cyou During training, I preserve the weights with consistent sizes while reinitialize the weights with inconsistent sizes(this operation is very common). I then gradually updated all the parameters and saved the weights in nyu. kept. During testing, I ultimately load the weights from nyu.ckpt, ensuring that I can obtain accurate results. It's my problem. I haven't organized the whole code yet. When initializating the PVT class, t the first step is to automatically load the pre-trained model of PVT. You can comment out https://github.com/WHU-USI3DV/SparseDC/blob/608cd1914fee3d816d25fdb1a9862058af759836/configs/experiment/final_version.yaml#L19 avoid warnings.

**Comment by lucky9-cyou** [comment]
> @lucky9-cyou During training, I preserve the weights with consistent sizes while reinitialize the weights with inconsistent sizes(this operation is very common). I then gradually updated all the parameters and saved the weights in nyu. kept. During testing, I ultimately load the weights from nyu.ckpt, ensuring that I can obtain accurate results. It's my problem. I haven't organized the whole code yet. When initializating the PVT class, the first step is to automatically load the pre-trained model of PVT. You can comment out
> 
> https://github.com/WHU-USI3DV/SparseDC/blob/608cd1914fee3d816d25fdb1a9862058af759836/configs/experiment/final_version.yaml#L19
> 
> to avoid warnings.

Got it. Thanks for your explanation. I will close this issue.
